File: 0017-letter-combinations-of-a-phone-number/depth-first-search.cs
public class Solution
{
    public void dfs(int idx, string digits, string[] map, StringBuilder tmp, IList<string> res)
    {
        if (idx == digits.Length)
        {
            res.Add(tmp.ToString());
            return;
        }
        else
        {
            var mapIndex = digits[idx] - '0';
            var letters = map[mapIndex];
            for (int i = 0; i < letters.Length; i++)
            {
                //Append character to combination
                tmp.Append(letters[i]);

                // Recursively call Depth First Search
                dfs(idx + 1, digits, map, tmp, res);

                // Remove last character from combination for backtracking
                tmp.Remove(tmp.Length - 1, 1);
            }
        }
    }

    public IList<string> LetterCombinations(string digits)
    {
        IList<string> result = new List<string>();
        var letters = new string[]{"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        var builder = new StringBuilder();

        if (digits.Length != 0)
        {
            dfs(0, digits, letters, builder, result);
        }

        return result;
    }
}

File: 0017-letter-combinations-of-a-phone-number/depth-first-search.rs
impl Solution
{
    pub fn depth_first_search(index: usize, digits: &String, map: &[&str; 8], tmp: &mut String, res: &mut Vec<String>)
    {
        if (index == digits.len())
        {
            res.push(tmp.clone());
            return;
        }
        else
        {
            // Build combination
            let numsString = map[((digits.chars().nth(index).unwrap() as u8) - b'2') as usize];

            for c in numsString.chars()
            {
                tmp.push(c);
                Self::depth_first_search(index + 1, digits, map, tmp, res);
                tmp.pop();
            }
        }
    }

    pub fn letter_combinations(digits: String) -> Vec<String>
    {
        let mut result: Vec<String> = Vec::new();
        let letters: [&str; 8] = ["abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"];

        if (digits.len() != 0)
        {
            let mut tree: String = String::new();
            Self::depth_first_search(0, &digits, &letters, &mut tree, &mut result);
        }

        return result;
    }
}